autoload/go/auto.vim

```diff
@@ -32,22 +32,46 @@ function! go#auto#echo_go_info()
   redraws! | echo "vim-go: " | echohl Function | echon item.info | echohl None
 endfunction
 
-function! go#auto#auto_type_info()
-  if !go#config#AutoTypeInfo() || !isdirectory(expand('%:p:h'))
-    return
-  endif
+function! go#auto#update_autocmd()
+  execute 'augroup vim-go-buffer-auto-' . bufnr('')
+    autocmd! * <buffer>
+    if go#config#AutoTypeInfo() || go#config#AutoSameids()
+      " vint: -ProhibitAutocmdWithNoGroup
+      autocmd CursorMoved <buffer> call go#auto#timer_restart()
+      autocmd BufLeave <buffer> call s:timer_stop()
+      " vint: +ProhibitAutocmdWithNoGroup
+    else
+      call s:timer_stop()
+    endif
+  augroup END
```

Comment:
Since this function will be executed on `CursorHold` and to make sure that duplicate events aren't configured, shouldn't it _first_ check to see whether there's an existing timer, and then
1. return early if there isn't an existing timer and neither auto type info nor nor auto same ids is enabled
1. return early when there is a timer and either one or both of auto type info and auto same ids is enabled
1. clear the group and stop the timer when there is a timer and neither auto type info nor auto same ids is enabled
1. configure the group when there isn't a timer and either one or both of auto type info and auto same ids is enabled

?

Reply:
Good point. It (re)defines augroup's redundantly. I fix it to do more effectively.

I arranged your suggestion in a table.

| # | has_timer | should_enable | do                                 |
|---|-----------|---------------|------------------------------------|
| 1 | false     | false         | return early                       |
| 2 | true      | true          | return early                       |
| 3 | true      | false         | clear the group and stop the timer |
| 4 | false     | true          | configure the group                |

Fixed on 1a5ff26